### PR TITLE
File.Unix: Replace: increase Windows compatibility

### LINKED ISSUE #49796: Unify File.Replace Unix and Windows exception throwing behavior
### Description

```cs
using (File.Create("test.txt")) {}
Directory.CreateDirectory("test");
			
// Correct - throws IOExeption with message "Is a directory"
File.Replace("test.txt", "test", null);

// Incorrect = throws IOExeption with message "Not a directory"
File.Replace("test", "test.txt", null);
```

### Configuration

.NET 5 on Linux.

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/runtime/issues/49796

- When source and destination are not a file throw UnauthorizedAccessException.
- When source and destination are the same file throw IOException.

@danmoseley @stephentoub ptal

cc @mikernet

### CODE REVIEW

File: src/libraries/System.IO.FileSystem/src/Resources/Strings.resx [hunk 1]
```diff
@@ -197,6 +197,9 @@
   <data name="IO_CannotCreateDirectory" xml:space="preserve">
     <value>The specified directory '{0}' cannot be created.</value>
   </data>
+  <data name="IO_CannotReplaceSameFile" xml:space="preserve">
+    <value>The source and destination are the same file.</value>
```
Comment (danmoseley) [inline src/libraries/System.IO.FileSystem/src/Resources/Strings.resx:201]: Include path? These are often in logs, where that can be helpful

**Comment by danmoseley** [comment]
This would need a breaking change note if we take it...

**Comment by danmoseley** [comment]
@marklio

**Comment by marklio** [comment]
Unfortunately, we don't have great ability to look at API usage by OS (we actively discourage those publishing patterns, particularly for managed libraries). All up, File.Replace is a widely-used API. Given its popularity and its age (around since NetFx 2.0), it is a safe bet that most of the code using this expects the semantics provided by Windows implementations. Since consistency is an important principle in .NET, consolidating on that behavior makes sense. 

It looks like this change affects more than exceptional scenarios (that is, things that didn't throw on Linux will now throw). If so, are there patterns that could be employed that would allow code to get the same replace semantics as the previous Linux behavior?

I also worry about the non-atomicity of the implemented checks. If these exceptional conditions are reliably atomic on Windows (the result of the underlying Win32 call), we've taken predictable (although inconsistent between OS) behavior and turned it into a potential reliability issue on Linux. I'm not a Linux filesystem expert. What would it take to provide atomicity here?

**Comment by danmoseley** [comment]
The atomicity only relates to whether you get the right exception. It doesn't affect whether the operation succeeds or fails.

We already have precedent for this:
https://github.com/dotnet/runtime/blob/1ee59da9f6104c611b137c9d14add04becefdf14/src/libraries/System.IO.FileSystem/src/System/IO/FileSystemInfo.Unix.cs#L69-L75

Unfortunately it's not avoidable I think if we want to throw the same exceptions.

File: src/libraries/System.IO.FileSystem/src/Resources/Strings.resx (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.IO.FileSystem/src/Resources/Strings.resx:201]: I am slowly working to add paths to all IO messages where we have a path available to include...

**Comment by mikernet** [comment]
FYI, the inconsistencies I've found in `File.Replace` are documented here:

https://github.com/Singulink/Singulink.IO.FileSystem/blob/master/Source/Singulink.IO.FileSystem/IAbsoluteFilePath.Impl.cs#L187

There are also inconsistencies in `File.Move`:

https://github.com/Singulink/Singulink.IO.FileSystem/blob/master/Source/Singulink.IO.FileSystem/IAbsoluteFilePath.Impl.cs#L187

There are also lots of other issues documented throughout that code file. The biggest one is the nonsense `UnauthorizedAccessException` messages that could probably be improved.

**Comment by tmds** [comment]
>  If so, are there patterns that could be employed that would allow code to get the same replace semantics as the previous Linux behavior?

`File.Move` and `Directory.Move` can be used.

> What would it take to provide atomicity here?

There's no way to check this atomically. The non-atomic checks will expose the cross-platform difference that are now only revealed when running on the other platform.

File: src/libraries/System.IO.FileSystem/src/Resources/Strings.resx (hunk 1 above)
Comment (tmds) [reply src/libraries/System.IO.FileSystem/src/Resources/Strings.resx:201]: I'll add the paths in the message.

**Comment by tmds** [comment]
> the inconsistencies I've found
> ...
> There are also lots of other issues 

@mikernet if there are some things you'd like, can you create issues for them? That will improve visibility, and there can be some discussion on what makes sense.

**Comment by tmds** [comment]
@danmoseley @marklio is this PR good to merge?

**Comment by danmoseley** [comment]
Paging this back in. The only behavior that used to work but now does not, is File.Copy(file, directory)` which currently works on Unix and after this will throw, is that a correct statement? 

I think the code change is fine.

**Comment by tmds** [comment]
> The only behavior that used to work but now does not, is File.Copy(file, directory) which currently works on Unix and after this will throw, is that a correct statement?

Yes, and copying from a file to the same file will now also throw.

**Comment by danmoseley** [comment]
@marklio 

> It looks like this change affects more than exceptional scenarios (that is, things that didn't throw on Linux will now throw). If so, are there patterns that could be employed that would allow code to get the same replace semantics as the previous Linux behavior?

The answer here is yes, on Unix you will need to replace `File.Copy(file, directory)` with `File.Copy(file, directory\file)`; and (much more minor) remove attempts to do `File.Copy(file, samefile)`.  Essentially the question is how do we feel about the first one.

**Comment by marklio** [comment]
> Essentially the question is how do we feel about the first one.

Agreed. It seems like one of our principles has been that carrying cross-plat behavioral differences into the future is untenable. I think we just need to be sure that the "correct" patterns are discoverable for folks moving forward that hit the exceptions. Usually, our tools there are breaking change docs. An article on the right way to author cross-plat file operations would probably be super useful.

**Comment by danmoseley** [comment]
@adamsitnik @carlossanlop @jozkee the IO crew can one of you please make a call and merge if you're comfortable.

**Comment by danmoseley** [comment]
@jkotas any objection to 

> File.Copy(file, directory) which currently works on Unix and after this will throw

**Comment by jkotas** [comment]
Fine with me.

> It seems like one of our principles has been that carrying cross-plat behavioral differences into the future is untenable.

We are trying to do a best effort Windows emulation on Unix here. I am pretty sure that there are still number of subtle differences, in particular around what is done atomically and what are possible error patterns.

Also, Windows emulation on Unix is not free. Notice that this change adds two extra file system syscalls to this APIs.

To write robust performant cross-platform code, one may want to avoid this type of APIs.

**Comment by jeffhandley** [comment]
@carlossanlop / @Jozkee / @adamsitnik -- Please target getting this reviewed so we can make a final decision before Preview 6 snaps.

**Review by stephentoub — APPROVED** [review]

**Comment by jeffhandley** [comment]
I resolved the conflicts. The `Strings.resx` additions needed to be manually moved over to `src/libraries/System.Private.CoreLib/src/Resources/Strings.resx` due to #53231. The other changes were automatically merged correctly.

**Comment by danmoseley** [comment]
`[ 1171.987738] .NET BGC[17401]: segfault at 7f2d472fbca0 ip 00007f2e3ed7ae0c sp 00007f2d8f7f9f80 error 4 in libcoreclr.so[7f2e3ea2f000+36b000]` running numerics tests.

Unfortunately no dump - our scripts may be faulty. Asking..

**Comment by jeffhandley** [comment]
I've restarted the checks and will check back in on them...

**Comment by danmoseley** [comment]
Thanks @tmds!

**Comment by carlossanlop** [comment]
Breaking change doc: https://github.com/dotnet/docs/issues/26323

**Pull request merged**